feat (toolbar):Debug toolbar

paulharter opened the pull request:
Checked out toolbar stuff into this new clean branch as many many messy changes of mind.

kevin-dp commented on .changeset/spicy-beans-build.md:
```diff
@@ -0,0 +1,6 @@
+---
+"@electric-sql/debug-toolbar": major
```
I believe this will bump the version of the toolbar to `1.0.0`. I think we want to keep it at `0.1.x` in which case this should be a patch bump.

paulharter replied:
The checkpoint tool wanted to set it to 1.0.0 as its a first version - couldn't see a way to do a patch

kevin-dp reviewed (changes requested):
The toolbar looks good overall, I left some comments.
Seems that we're close to be able to merge this 👍

kevin-dp commented:
Also, it seems you had an example usage of the toolbar in web-wa-sqlite example but you removed it?
Would be good to have an example where the toolbar is being used such that users can see the usage.

paulharter commented on pnpm-workspace.yaml:
```diff
@@ -4,3 +4,4 @@ packages:
   - 'examples/starter'
   - 'generator'
   - 'components/electric'
+  - 'toolbar'
```
I'm easy about this which ever seems best

paulharter commented on toolbar/.prettierignore:
```diff
@@ -0,0 +1,12 @@
+*.md
+.DS_Store
+dist/
+dev-app/dist/
```
I keep the dev-app locally for development so this is useful

paulharter commented on toolbar/README.md:
```diff
@@ -0,0 +1,92 @@
+// toolbar imports
+import { globalRegistry } from "electric-sql/satellite";
+import addToolbar, { typescriptApi } from '@electric-sql/debug-toolbar'
+import '@electric-sql/debug-toolbar/dist/index.cjs.css'
```
I couldn't get it to work otherwise - what would be the better way?

paulharter commented on toolbar/README.md:
```diff
@@ -0,0 +1,92 @@
+You can use one of the example projects such as `examples/web-wa-sqlite`, 
+but you will have to make a few changes to the project's configuration:
+
+Add web-wa-sqlite to the pnpm workspace by adding this line to `pnpm-workspace.yaml`
```
I've removed this section on development

paulharter commented on toolbar/tsconfig.json:
```diff
@@ -0,0 +1,27 @@
+    "strict": true, // Enable all strict type checking options
+    "target": "ESNext" // Specify ECMAScript target version
+  },
+  "include": ["src/*", "custom.d.ts"]
```
I'm not sure - I'll try

paulharter commented on toolbar/src/api/toolbar-typescript.ts:
```diff
@@ -0,0 +1,36 @@
+
+  queryDB(dbName: string, statement: Statement): Promise<Row[]> {
+    const sat = this.globalRegistry.satellites[dbName]
+    return sat?.adapter.query(statement) ?? Promise.resolve([])
```
I've replaced this with an error message

paulharter commented on toolbar/src/index.tsx:
```diff
@@ -0,0 +1,80 @@
+
+function ElectricToolbar({ api }: ToolbarProps) {
+  const [hidden, setHidden] = useState(true)
+  const [dbNames, setDbNames] = useState([''])
```
React seems to want to know the type of the useState variable

paulharter commented on toolbar/src/tabs.tsx:
```diff
@@ -0,0 +1,72 @@
+
+  function renderComp() {
+    switch (active) {
+      case 'status':
```
I had something like this before but it upset React - It give hook errors something to do with when the JSX objects get rendered

paulharter replied:
I've removed the status one

paulharter commented on toolbar/src/tabs/SQLTab.tsx:
```diff
@@ -0,0 +1,135 @@
+    })
+  }
+
+  function clearResults() {
```
makes sense

paulharter commented on toolbar/src/tabs/SQLTab.tsx:
```diff
@@ -0,0 +1,135 @@
+              <span className="header-span">query</span>
+              <span
+                className="header-span header-span-button"
+                onClick={() => {
```
same

paulharter commented on toolbar/src/tabs/SQLTab.tsx:
```diff
@@ -0,0 +1,135 @@
+              <span
+                className="header-span header-span-button"
+                onClick={() => {
+                  setActive('query')
```
yup

paulharter commented on toolbar/src/tabs/SQLTab.tsx:
```diff
@@ -0,0 +1,135 @@
+                  setCode(value)
+                }}
+                options={{
+                  tabSize: 2,
```
I've set it to 4 everywhere

paulharter commented on toolbar/src/tabs/SQLTab.tsx:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  function switchInput() {
+    switch (active) {
```
done

paulharter commented on toolbar/src/tabs/StatusTab.tsx:
```diff
@@ -0,0 +1,23 @@
+  const [status, setStatus] = useState('')
+
+  useEffect(() => {
+    if (dbName !== undefined) {
```
done

paulharter commented on clients/typescript/package.json:
```diff
@@ -60,6 +60,8 @@
     "./wa-sqlite": "./dist/drivers/wa-sqlite/index.js",
     "./config": "./dist/config/index.js",
     "./debug": "./dist/util/debug/index.js",
+    "./satellite/mock": "./dist/satellite/mock.js",
```
I played with this - I added `export { MockRegistry } from './mock'` to `satellite/index.ts` to make it work

kevin-dp replied:
yep that's the way to do it 👍

paulharter commented:
@kevin-dp I've pretty done everything as you suggest. A few notes:

I had to add export { MockRegistry } from './mock' to satellite/index.ts to get the import to work without the addition export is the package

I couldn't refactor the tab components as you suggested, I had tried to do this before, its something to do with when react expands out the JSX, it generates react hook errors.

I have moved the whole thing for now - but we can do this later

I haven't added back the toolbar in an example as I want to do this an other step once there is a published toolbar package to refer to properly. But I have checked that this will works ok in the web example. I have also added a non compiled `util.ts` script to tidy away the code to add the toolbar.

kevin-dp commented on toolbar/src/index.tsx:
```diff
@@ -0,0 +1,82 @@
+  return new ToolbarTypescript(globalRegistry)
+}
+
+export default function addToolbar(toolbarApi: ToolbarInterface) {
```
I would not use a default export here because it makes the code that imports it less clear:
```ts
toolbar.default(toolbar.clientApi(...))
```
In the code above, it's not clear what `toolbar.default` does.
I would argue that this is more clear:
```ts
toolbar.addToolbar(toolbar.clientApi(...))
```

And if using regular imports instead of dynamic imports it becomes even better:
```ts
import { addToolbar, clientApi } from '...'
addToolbar(clientApi(...))
```

paulharter replied:
removed all the dynamic imports

kevin-dp commented on toolbar/util.ts:
```diff
@@ -0,0 +1,9 @@
+export function addToolbar() {
```
Let's not use dynamic import and instead have `addToolbar` take 1 argument which is satellite or electric.

paulharter replied:
I couldn't get this to work - the typing of the client makes it want to pulll in generated types

so I left it with `addToolbar(globalRegistry)`

kevin-dp replied:
Try `ElectricClient<unknown>` otherwise you could do `ElectricClient<any>` that should make TS happy.

paulharter commented on toolbar/package.json:
```diff
@@ -0,0 +1,42 @@
+  "author": "ElectricSQL",
+  "license": "Apache-2.0",
+  "scripts": {
+    "build": "rm -rf ./dist && node builder.js && tsc --emitDeclarationOnly --outDir dist && cp ./util.ts ./dist/util.ts",
```
I've removed this in favour of static imports

paulharter commented on toolbar/README.md:
````diff
@@ -0,0 +1,47 @@
+In your code add the toolbar after initialising the electric database like this.
+
+```typescript
+import ('@electric-sql/debug-toolbar/util').then(util => {util.addToolbar()});
````
I have removed all the dynamic imports

paulharter commented on toolbar/src/api/toolbar-interface.ts:
```diff
@@ -0,0 +1,8 @@
+export interface ToolbarInterface {
+  getSatelliteNames(): string[]
+  getSatelliteStatus(name: string): ConnectivityState | 'Not found'
+  resetDB(dbName: string): void
```
done

kevin-dp commented on toolbar/src/api/toolbar-typescript.ts:
```diff
@@ -0,0 +1,40 @@
+import { Row, Statement, ConnectivityState } from 'electric-sql/util'
+import { GlobalRegistry } from 'electric-sql/satellite'
+
+export class ToolbarTypescript implements ToolbarInterface {
```
This is a typescript package so no need to put `Typescript` in the class name nor in the filename.
Can we rename this file to `toolbar.ts` and this class to `Toolbar` please. Also, we can rename the `toolbar-interface.ts` file `interface.ts`

paulharter replied:
I was anticipating that this will could be used with the rust client too - but happy to remove the language specification if you prefer.

kevin-dp replied:
Yes please :-)

paulharter replied:
@kevin-dp have changed as you suggest now

kevin-dp commented on toolbar/README.md:
````diff
@@ -32,14 +32,14 @@ First some imports
 ```
 
-Then in your code add the toolbar after initialising the electric database like this:
+Then in your code after calling `electrify` pass the electric client into addToolbar:
````
addToolbar should be `addToolbar`

kevin-dp commented on toolbar/README.md:
````diff
@@ -32,14 +32,14 @@ First some imports
 
 ```typescript
-addToolbar(globalRegistry)
+const electricClient = await electrify(conn, schema, config)
````
we usually just call it `electric`

paulharter commented:
I think this might be ready now - I've made to readme changes you suggested

kevin-dp commented on .github/workflows/toolbar_tests.yml:
```diff
@@ -0,0 +1,63 @@
+          node-version: 18
+          cache: pnpm
+      - run: make deps
+      - run: pnpm run build
```
I don't see the change? i still see `pnpm run build`

kevin-dp commented on clients/typescript/src/satellite/index.ts:
```diff
@@ -27,6 +27,7 @@ import {
 } from './shapes/types'
 import { ShapeSubscription } from './process'
 import { DbSchema } from '../client/model/schema'
+export { MockRegistry } from './mock'
```
Could you move it down to the block of exports.
It's in the block of imports which confused me, i first thought it was an import.

kevin-dp commented on components/toolbar/.gitignore:
```diff
@@ -0,0 +1,7 @@
+dist
+node_modules
+dev-app
```
`dev-app` and `package.dev.json` are not needed. You have them locally because of the way how you tested this but normally one doesn't have this directory and file so no need to put it in `.gitignore`.

kevin-dp commented on components/toolbar/.prettierignore:
```diff
@@ -0,0 +1,8 @@
+node_modules/
+.gitignore
+.github/
+dev-app/src/generated/
```
same here, `dev-app` is not needed.

kevin-dp commented on components/toolbar/Makefile:
```diff
@@ -0,0 +1,9 @@
+deps:
+	pnpm install --frozen-lockfile
```
indentation is off, we usually indent it like this for targets containing several commands:
```
deps:
  pnpm install --frozen-lockfile
  make -C ...
```

And like this for targets with a single command:
```
tests: CI=true pnpm run test
```

kevin-dp commented on components/toolbar/src/api/interface.ts:
```diff
@@ -0,0 +1,8 @@
+import { Row, Statement, ConnectivityState } from 'electric-sql/util'
+
+export interface Interface {
```
Can we give this a more descriptive name?
I would be fine with `ToolbarInterface`, `IToolbar`, or other descriptive names.

kevin-dp commented on components/toolbar/src/index.tsx:
```diff
@@ -0,0 +1,85 @@
+  const [dbName, setDbName] = useState('')
+
+  useEffect(() => {
+    let names = api.getSatelliteNames()
```
We don't reassign `names` so better to use `const`.

kevin-dp commented on components/toolbar/src/index.tsx:
```diff
@@ -0,0 +1,85 @@
+
+export function addToolbar(electric: ElectricClient<any>) {
+  const toolbarApi = clientApi(electric.registry)
+  const toolbar_div = document.createElement('div')
```
we use camelCase in TS -> `toolbarDiv`

kevin-dp commented on components/toolbar/src/index.tsx:
```diff
@@ -0,0 +1,85 @@
+  toolbar_div.setAttribute('id', 'electric-toolbar')
+  toolbar_div.setAttribute('class', 'electric-toolbar')
+  document.body.appendChild(toolbar_div)
+  const toolbar_root = ReactDOM.createRoot(
```
`toolbarRoot`

kevin-dp commented on components/toolbar/src/tabs.tsx:
```diff
@@ -0,0 +1,66 @@
+
+function TabItem(
+  label: string,
+  name: 'status' | 'db' | 'sql',
```
this type is being used in several places.
Let's make a type alias for it.

kevin-dp commented on components/toolbar/src/tabs.tsx:
```diff
@@ -0,0 +1,66 @@
+  api,
+}: ToolbarTabsProps): JSX.Element {
+  const [active, setActive] = useState<'status' | 'db' | 'sql'>('status')
+  function handleClick(
```
`handleClick` does not use the event and just calls `setActive`. Do we really need `handleClick`? It seems like we can get rid of this function and just pass `secActive` into `TabItem` if we change the type of `handleClick` in `TabItem` to:
```ts
handleClick: (
    name: 'status' | 'db' | 'sql'
  ) => void
```

paulharter replied:
neat - I wasn't sure this would work

kevin-dp commented on components/toolbar/src/tabs/SQLTab.tsx:
```diff
@@ -0,0 +1,140 @@
+  const [active, setActive] = useState('query')
+
+  function submitSQL() {
+    let cmd = code
```
Why is this variable needed, `cmd` is not reassigned. Could use `code` instead of `cmd`.

kevin-dp commented on components/toolbar/test/api.test.ts:
```diff
@@ -0,0 +1,51 @@
+import { DbSchema } from 'electric-sql/client/model'
+
+test('identity', async () => {
+  const dbDescription = {} as DbSchema<any>
```
Could type the variable such that we don't have to cast the value `const dbDescription: DbSchema<any> = {}`

paulharter replied:
I get an error if I do this - {} isn't actually sufficient to be a DbSchema<any> so I'm forcing it

kevin-dp replied:
I see, we can leave it like this then

kevin-dp commented on components/toolbar/test/api.test.ts:
```diff
@@ -0,0 +1,51 @@
+import { clientApi } from '../src'
+import { DbSchema } from 'electric-sql/client/model'
+
+test('identity', async () => {
```
this tests `getSatelliteNames`, we should also test the other public methods:
```ts
getSatelliteStatus(name: string): ConnectivityState | 'Not found'
resetDB(dbName: string): Promise<void>
queryDB(dbName: string, statement: Statement): Promise<Row[]>
```

balegas commented:
I see there is still activity here. Let's not leave it hanging.

paulharter commented:
@kevin-dp I have done all the changes you suggested except the additional tests, and I have tested it manually against 0.0.9 with the web example.

Can we merge this now as I'm not really coding now and don't have time to think about how to set up the extra tests and Valter wanted this in for next week?

kevin-dp commented on components/toolbar/src/tabs.tsx:
```diff
@@ -0,0 +1,59 @@
+function TabItem(
+  label: string,
+  name: TabName,
+  handleClick: (name: 'status' | 'db' | 'sql') => void,
```
you can use `TabName` here as the type for `name`

paulharter replied:
done

kevin-dp reviewed (commented):
Thanks for addressing my comments!
I have two nitpicks that would be nice to fix:
- `components/toolbar/Makefile` indentation is still off
- `clients/typescript/src/satellite/index.ts` export is still on the block of imports

You tested the toolbar manually on 0.9 and it seems to work which is great. However, we don't have unit tests for some public methods and that would be a blocker for me as that means we will have to go and manually test the toolbar on every release. These are the untested public methods:
```ts
getSatelliteStatus(name: string): ConnectivityState | 'Not found'
resetDB(dbName: string): Promise<void>
queryDB(dbName: string, statement: Statement): Promise<Row[]>
```

I will let @balegas decide whether we merge without the tests or not.

balegas commented:
Hi @paulharter,
It feels like the unit tests suggestion are important. I understand that this will probably not help a lot with browser implementation, but for the sake of completeness let's make those tests.

This PR is dragging for too long. It's very near.

Best,

paulharter commented:
@kevin-dp @balegas I've added the extra api tests now

msfstef commented:
I've modified the debug toolbar slightly, added a couple more APIs for the connection status (toggle on/off) and an experimental shapes tab for viewing active shapes.

The last one uses a private field from the satellite implementation so before developing further I'll open a PR to expose some shape subscription information from the Satellite client for the debug toolbar to use safely.

I've also done an ugly inlining of CSS into the debug toolbar to make the DX better (now only need to dynamically import and call `addToolbar(electric)`) - will improve that over time but I had already spent too much time on CSS bunding, postprocessing, and TS errors...

@kevin-dp can you have another look so we can potentially merge this?

kevin-dp commented on components/toolbar/package.json:
```diff
@@ -0,0 +1,47 @@
+    "prettier": "3.2.5",
+    "typescript": "^5.4.5",
+    "vitest": "^1.5.0",
+    "zod": "^3.23.3"
```
Does the toolbar need Zod?
If yes, this version is problematic as it may give type errors in the generated client, let's use `3.21.1` instead.

kevin-dp reviewed (approved):
Looks good to me!
Left some minor comments.

msfstef commented on components/toolbar/Makefile:
```diff
@@ -0,0 +1,9 @@
+deps:
```
Not sure what you mean? it's tab indented (similar to the generator Makefile)
